SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/src/ServerlessDocAnalysis/StepFunctions/StepFunctionConstruct.cs

```diff
@@ -0,0 +1,193 @@
+            Comment = "Function to send document to textract asynchronously for expense analysis",
+            Payload = TaskInput.FromObject(new Dictionary<string, object> {
+            { "id", JsonPath.StringAt("$.id") },
+            { "taskToken", JsonPath.TaskToken}
+            })
+        });
+
+        LambdaInvoke processTextractExpenseResultsState = new(this, "processTextractExpenseResults", new LambdaInvokeProps
+        {
+            LambdaFunction = props.ProcessTextractExpenseFunction,
+            Comment = "Function to process textract Expense results asynchronously",
+            OutputPath = "$.Payload",
+        });
+
+        SqsSendMessage sendFailureQueue = new(this, "sendFailureQueue", new SqsSendMessageProps
+        {
+            Queue = props.SendFailureQueue,
+            Comment = "Send Failure Message",
+            MessageBody = TaskInput.FromJsonPathAt("$"),
+        });
+
+        SqsSendMessage sendSuccessQueue = new(this, "sendSuccessQueue", new SqsSendMessageProps
+        {
+            Queue = props.SendSuccessQueue,
+            Comment = "Send Success Message",
+            MessageBody = TaskInput.FromJsonPathAt("$")
+        });
+
+        // Compose the workflow sequence
+        initializeState.Next(textractState);
+        initializeState.AddCatch(sendFailure, new CatchProps
+        {
+            Errors = new[] { "States.ALL" },
+            ResultPath = "$.error"
+        });
+        textractState.Next(processTextractResultsState);
+        textractState.AddCatch(sendFailure, new CatchProps
+        {
+            Errors = new[] { "States.ALL" },
+            ResultPath = "$.error"
+        });
+
+        processTextractResultsState.Next(textractExpenseState);
+        processTextractResultsState.AddCatch(sendFailure, new CatchProps
+        {
+            Errors = new[] { "States.ALL" },
+            ResultPath = "$.error"
+        });
+
+        textractExpenseState.Next(processTextractExpenseResultsState);
+        textractExpenseState.AddCatch(sendFailure, new CatchProps
+        {
+            Errors = new[] { "States.ALL" },
+            ResultPath = "$.error"
+        });
+
+        processTextractExpenseResultsState.Next(sendSuccess);
+        processTextractExpenseResultsState.AddCatch(sendFailure, new CatchProps
+        {
+            Errors = new[] { "States.ALL" },
+            ResultPath = "$.error"
+        });
+
+        sendSuccess.Next(sendSuccessQueue);
+
+        sendFailure.Next(sendFailureQueue);
+
+
+        // Log group for the step function
+        LogGroup stepFunctionLogGroup = new(this, "stepFunctionLogGroup", new Amazon.CDK.AWS.Logs.LogGroupProps
+        {
+            LogGroupName = GetLogGroupName("docProcessingWorkflow"),
+            RemovalPolicy = RemovalPolicy.DESTROY
+        });
+
+        // Update the step function body and use the passed in props
+        StateMachine = new StateMachine(this, $"{id}StateMachine", new StateMachineProps
```

Comment:
Can extract different parts of build to methods and only leave on level of abstraction - https://www.code4it.dev/cleancodetips/coherent-levels-of-abstraction/

Reply:
Abstracted the "build workflow chain structure" to another method. Left definition of different step / Log / EB resources in the constructer (to resemble typical CDK structure). Should make it look quite a bit simpler and easier to follow.